File: RaceControlApp/RaceControl/Models/Downsample.swift
import Foundation

/// Largest-Triangle-Three-Buckets downsampling.
///
/// A telemetry trace contains thousands of samples per channel. LTTB preserves
/// visually significant points such as braking peaks while capping the number
/// of marks handed to Swift Charts. This mirrors Android's `Downsample` helper
/// so both clients make the same trade-off.
enum Downsample {

    static func lttb(
        xs: [Double],
        ys: [Double],
        threshold: Int
    ) -> (xs: [Double], ys: [Double]) {
        let count = min(xs.count, ys.count)
        guard threshold < count, threshold >= 3, count >= 3 else {
            return (Array(xs.prefix(count)), Array(ys.prefix(count)))
        }

        var outputX: [Double] = []
        var outputY: [Double] = []
        outputX.reserveCapacity(threshold)
        outputY.reserveCapacity(threshold)

        let bucketWidth = Double(count - 2) / Double(threshold - 2)
        var selectedIndex = 0
        outputX.append(xs[0])
        outputY.append(ys[0])

        for bucket in 0..<(threshold - 2) {
            let averageStart = Int(Double(bucket + 1) * bucketWidth) + 1
            let averageEnd = min(Int(Double(bucket + 2) * bucketWidth) + 1, count)
            let averageCount = max(averageEnd - averageStart, 1)

            var averageX = 0.0
            var averageY = 0.0
            if averageStart < averageEnd {
                for index in averageStart..<averageEnd {
                    averageX += xs[index]
                    averageY += ys[index]
                }
            }
            averageX /= Double(averageCount)
            averageY /= Double(averageCount)

            let rangeStart = Int(Double(bucket) * bucketWidth) + 1
            let rangeEnd = min(Int(Double(bucket + 1) * bucketWidth) + 1, count)
            let pointAX = xs[selectedIndex]
            let pointAY = ys[selectedIndex]

            var maximumArea = -1.0
            var maximumIndex = rangeStart
            if rangeStart < rangeEnd {
                for index in rangeStart..<rangeEnd {
                    let area = abs(
                        (pointAX - averageX) * (ys[index] - pointAY)
                            - (pointAX - xs[index]) * (averageY - pointAY)
                    ) * 0.5
                    if area > maximumArea {
                        maximumArea = area
                        maximumIndex = index
                    }
                }
            }

            outputX.append(xs[maximumIndex])
            outputY.append(ys[maximumIndex])
            selectedIndex = maximumIndex
        }

        outputX.append(xs[count - 1])
        outputY.append(ys[count - 1])
        return (outputX, outputY)
    }

    static func cap(
        xs: [Double],
        ys: [Double],
        maximum: Int = 800
    ) -> (xs: [Double], ys: [Double]) {
        lttb(xs: xs, ys: ys, threshold: maximum)
    }
}

File: RaceControlAndroid/app/src/main/java/com/owlmedia/racecontrol/core/util/Downsample.kt
package com.owlmedia.racecontrol.core.util

import androidx.compose.foundation.layout.size

/**
 * Largest-Triangle-Three-Buckets downsampling.
 *
 * A fastest-lap telemetry trace is several thousand samples per channel, and
 * handing all of them to a chart is the single easiest way to make this app
 * drop frames on a mid-range phone. LTTB keeps the visually significant points
 * (peaks, braking edges) rather than naively taking every Nth sample, which
 * would smooth away exactly the detail the chart exists to show.
 */
object Downsample {

    fun lttb(xs: List<Double>, ys: List<Double>, threshold: Int): Pair<List<Double>, List<Double>> {
        val n = minOf(xs.size, ys.size)
        if (threshold >= n || threshold < 3 || n < 3) {
            return xs.take(n) to ys.take(n)
        }

        val outX = ArrayList<Double>(threshold)
        val outY = ArrayList<Double>(threshold)

        // Buckets between the mandatory first and last points.
        val every = (n - 2).toDouble() / (threshold - 2)

        var a = 0
        outX.add(xs[0]); outY.add(ys[0])

        for (i in 0 until threshold - 2) {
            // Average of the *next* bucket, used as the triangle's third vertex.
            val avgRangeStart = ((i + 1) * every).toInt() + 1
            var avgRangeEnd = ((i + 2) * every).toInt() + 1
            if (avgRangeEnd > n) avgRangeEnd = n

            var avgX = 0.0
            var avgY = 0.0
            val avgCount = (avgRangeEnd - avgRangeStart).coerceAtLeast(1)
            for (j in avgRangeStart until avgRangeEnd) {
                avgX += xs[j]
                avgY += ys[j]
            }
            avgX /= avgCount
            avgY /= avgCount

            val rangeOffs = (i * every).toInt() + 1
            var rangeTo = ((i + 1) * every).toInt() + 1
            if (rangeTo > n) rangeTo = n

            val pointAX = xs[a]
            val pointAY = ys[a]

            var maxArea = -1.0
            var maxAreaIndex = rangeOffs
            for (j in rangeOffs until rangeTo) {
                val area = kotlin.math.abs(
                    (pointAX - avgX) * (ys[j] - pointAY) - (pointAX - xs[j]) * (avgY - pointAY)
                ) * 0.5
                if (area > maxArea) {
                    maxArea = area
                    maxAreaIndex = j
                }
            }

            outX.add(xs[maxAreaIndex])
            outY.add(ys[maxAreaIndex])
            a = maxAreaIndex
        }

        outX.add(xs[n - 1]); outY.add(ys[n - 1])
        return outX to outY
    }

    /** Convenience for the common "cap a telemetry channel" case. */
    fun cap(xs: List<Double>, ys: List<Double>, max: Int = 800): Pair<List<Double>, List<Double>> =
        lttb(xs, ys, max)
}
